### ISSUE TITLE
What should the return value from m.Get be?

### ISSUE THREAD

**Issue by arlolra** [coniks-sys/libmerkleprefixtree-go#10]
Moving this discussion from #1 so it doesn't get lost.

---

The idea was that `Get` would return a `MerkleNode`.  And then you'd have another function,

```
func GenerateProof(mn MerkleNode) []ProofNode {
  var proof []ProofNode
  for mn.parent !== nil {
    proof = preprend(...)
    mn = mn.parent
  }
  return proof
}
```

@masomel said,

> I think it makes sense to separate the Get from the proof generation logic, especially if the lookup returns an EmptyNode whenever the lookup failed, as long as there is still a lookup function that returns the proof node. As we're using the Merkle tree as an authenticated dictionary, the lookups should always return the path along with the value Section 3 [1].

[1] https://dl.acm.org/citation.cfm?doid=2019599.2019602

**Comment by c633** [comment]
This may include #7

**Comment by c633** [comment]
If we reached a different leaf with a matching prefix, what's `Get` should return? It cannot return the `userLeafNode` that has matching prefix. So if it returns `nil`, we cannot construct the auth path. 

I think `Get` would return a `MerkleNode` (or `nil`) as current, along with _a_ `ProofNode` which is an `emptyNode` or an `userLeafNode`. Then we'd have `GenerateProof` function that takes the returned `ProofNode` as the parameter.

```
func GenerateProof(mn ProofNode) []ProofNode {
  var proof []ProofNode
  for mn.parent !== nil {
    proof = prepend(...)
    mn = mn.parent
  }
  return proof
}
```

**Comment by masomel** [comment]
I'm honestly a bit confused about why we need ```MerkleNode``` and ```ProofNode``` (and I might just be a bit slow because I'm still learning Go). ```MerkleNode``` is an interface that implements common methods for our three node types, right? ```ProofNode``` is an interface to a node's hash?

The reason why I'm confused is because I don't really understand why these are separated. I'm fine with ```Get``` returning a ```MerkleNode```if what the means is that it'll either return a ```userLeafNode``` or an ```EmptyNode```. But I also think that ```Get``` should return the proof. Merkle tree lookups should return the authentication path by definition, but also because ```Get``` is already traversing the tree, so why have a separate method that traverses the tree again to just generate the proof?

Now, remember that the proof consists of a path in the tree, from the leaf to the root, and all the interior nodes (including the root) are pruned. I currently don't see where the tree is actually being pruned. The hashes of the path should be cleared as these will be recomputed by the verifier, and the pointers to the pruned subtrees should be set to ```nil```. If I've understood ```node```'s ```GetHash()``` correctly, right now it returns the hash of a node by returning the parent node's left or right child hash, depending on whether the ```node``` is the left or right child. So I don't really understand what ```ProofNode``` is being used for if it doesn't really represent a pruned node.

**Comment by c633** [comment]
> I'm honestly a bit confused about why we need MerkleNode and ProofNode (and I might just be a bit slow because I'm still learning Go). MerkleNode is an interface that implements common methods for our three node types, right? ProofNode is an interface to a node's hash?

My idea is to use `MerkleNode` interface to access the `value` of the returned node (if it is an `userLeafNode`), whereas the `ProofNode` just provides methods to access some other values (e.g., the full index of the node in auth path). If we use `MerkleNode` instead of `ProofNode` in the returned auth path, there are chances for users to read the `value` of the leaf node in the path, and this is what we want to avoid, right?

> But I also think that Get should return the proof. 

But you also said this before:
> I think it makes sense to separate the Get from the proof generation logic.

Did I misunderstand your meaning here?

**Comment by masomel** [comment]
> But you also said this before:
>> I think it makes sense to separate the ```Get``` from the proof generation logic.

I'm sorry. I take this back. I wasn't thinking about the fact that separating the two duplicates the tree traversal.

> My idea is to use ```MerkleNode``` interface to access the ```value``` of the returned node (if it is an ```userLeafNode```), whereas the ```ProofNode``` just provides methods to access some other values (e.g., the full index of the node in auth path). If we use ```MerkleNode``` instead of ```ProofNode``` in the returned auth path, there are chances for users to read the value of the leaf node in the path, and this is what we want to avoid, right?

Right. What I was misunderstanding is what access the users have to the underlying ```node```'s fields if they're using an interface versus the struct. So you use ```MerkleNode``` instead of exposing the ```value``` of a ```userLeafNode``` directly since the user would have access to it, even when they aren't supposed to in a proof of absence, right? I think I see now. Thanks for bearing with me.

So if the ```ProofNode``` exposes the ```index``` and the hash of the pruned subtree, right? It still looks to me like that's not what the current implementations of ```GetHash``` are doing.

**Comment by c633** [comment]
> Now, remember that the proof consists of a path in the tree, from the leaf to the root, and all the interior nodes (including the root) are pruned. I currently don't see where the tree is actually being pruned. The hashes of the path should be cleared as these will be recomputed by the verifier, and the pointers to the pruned subtrees should be set to nil. If I've understood node's GetHash() correctly, right now it returns the hash of a node by returning the parent node's left or right child hash, depending on whether the node is the left or right child. So I don't really understand what ProofNode is being used for if it doesn't really represent a pruned node.

Yup, thanks for making it clear. I understand it now.

**Comment by arlolra** [comment]
> If we use MerkleNode instead of ProofNode in the returned auth path, there are chances for users to read the value of the leaf node in the path, and this is what we want to avoid, right?

By "users" here, as in the email thread, I'm assuming you mean "developers"?  But they can just cast to whatever they want. In any case, we established in https://github.com/coniks-sys/libmerkleprefixtree-go/pull/1#discussion_r67771848 that the `value` needs to be encrypted to afford any privacy from the server. So, let's not worry about that.

I think you both asked very pertinent questions here:

> If we reached a different leaf with a matching prefix, what's Get should return?

> Get is already traversing the tree, so why have a separate method that traverses the tree again to just generate the proof?

The reason for this issue was because I was wondering just how general the Merkle tree should be.  I guess I imagined a `Get` that returned whatever node it found and then some wrapping function that might compare the indices and return `nil` if not matched, and also generate the authentication path.  The wrapping function implementing the specifics for our use of the tree.

But, in light of the reference above, I should probably just think of the Merkle tree implementation as a PAD, and not worry about overloading it.

Sorry for the confusion.
